### PR TITLE
POC for multi-session app

### PR DESCRIPTION
**Pull request by NSeydoux**
**This should not be merged in master, it is just an experiment. **

This code demonstrates what multi-session management could look like without an actual `SessionManager`, but rather a `getStoredSession` function.

### CODE REVIEW

File: packages/node/example/multiSession/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Demo: a server NodeJS app managing sessions
```
Comment (pmcb55) [inline packages/node/example/multiSession/README.md:1]:
```suggestion
# Demo: a server-side NodeJS application managing multiple sessions
```

File: packages/node/example/multiSession/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+This demo shows how you can run a NodeJS server with multiple sessions in parallel.
+Check out the code in `src/serverSideApp.ts` for more details.
+
+Note that until [this issue](https://github.com/solid/node-solid-server/issues/1533)
+is resolved, this will not work against NSS. This demo is only intended to work against a compliant
+OIDC implementation, such as Inrupt's ESS (e.g. deployed at `https://pod.inrupt.com`).
```
Comment (pmcb55) [inline packages/node/example/multiSession/README.md:8]: Our writing guidelines state 'i.e.' and 'e.g.' should be followed by a comma. I'd never of heard of it until Kevin brought it up recently (and it still seems weird to me), but apparently it is more correct English.
```suggestion
OIDC implementation, such as Inrupt's ESS (e.g., deployed at `https://pod.inrupt.com`).
```

File: packages/node/example/multiSession/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+## Installing the app
+
+To properly install this demo app, first run `npm ci` at the root of the repository, which will trigger
+the build of the library. Then, run `npm ci` again, this time in
+`packages/node/example/multiSession/`, and then `npm run build`. You should be all set. If you change the
```
Comment (pmcb55) [inline packages/node/example/multiSession/README.md:14]:
```suggestion
To install this demo app, first run `npm ci` at the root of the repository, which will trigger
the build of the libraries our app depends on. Then, run `npm ci` again, this time in
`packages/node/example/multiSession/`, and then `npm run build`. You should be all set. If you change the
```

File: packages/node/example/multiSession/README.md (hunk 2 above)
Comment (NSeydoux) [reply packages/node/example/multiSession/README.md:8]: Oh right, yeah I was told that rule in the past

File: packages/node/example/multiSession/src/fileSystemStorage.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Simple persistent storage solution that uses a file on the host filesystem to store data
+ * @hidden
+ */
+export class FileSystemStorage implements IStorage {
+  private map: Record<string, string> = {};
+
+  constructor(private path: string) {
+    // If the file does not exist, create it
```
Comment (pmcb55) [inline packages/node/example/multiSession/src/fileSystemStorage.ts:41]:
```suggestion
    // If the file does not exist, create it.
```

File: packages/node/example/multiSession/src/serverSideApp.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+const app = express();
+const PORT = 3001;
+const storage = new FileSystemStorage("sessions-data.json");
+
+const DEFAULT_OIDC_ISSUER = "https://broker.pod.inrupt.com/";
+
+app.use(
+  cookieSession({
+    name: "session",
+    keys: ["let me tell you a secret", "love isn't an actual ingredient"],
```
Comment (pmcb55) [inline packages/node/example/multiSession/src/serverSideApp.ts:39]: As I said last week, I'd just use text here explaining that this `key` is mandatory for cookieSession, but not actually necessary or used in our little app.

File: packages/node/example/multiSession/src/serverSideApp.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      res.writeHead(302, {
+        location: data,
+      });
+      res.end();
+    },
+  });
+  if (session.info.isLoggedIn) {
+    res
+      .writeHead(200, { "Content-Type": "text/html" })
+      .write(`<p>Alredy logged in</p>`);
```
Comment (pmcb55) [inline packages/node/example/multiSession/src/serverSideApp.ts:61]:
```suggestion
      .write(`<p>Already logged in as [${session.info.webId}]</p>`);
```

File: packages/node/example/multiSession/src/serverSideApp.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+app.get("/redirect", async (req, res) => {
+  const session =
+    (await getStoredSession(req.session!.sessionId, storage)) ??
+    new Session({ storage }, req.session!.sessionId);
+
+  await session.handleIncomingRedirect(getRequestFullUrl(req));
+  if (session.info.isLoggedIn) {
+    res
+      .writeHead(200, { "Content-Type": "text/html" })
+      .write(`<p>Logged in as ${session.info.webId} after redirect</p>`);
```
Comment (pmcb55) [inline packages/node/example/multiSession/src/serverSideApp.ts:75]:
```suggestion
      .write(`<p>Logged in as [${session.info.webId}] after redirect</p>`);
```

File: packages/node/example/multiSession/src/serverSideApp.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  res.end();
+});
+
+app.get("/logout", async (req, res, next) => {
+  const session = await getStoredSession(req.session!.sessionId, storage);
+  if (session) {
+    session.logout();
+    res
+      .writeHead(200, { "Content-Type": "text/html" })
+      .write(`<p>logged out</p>`);
```
Comment (pmcb55) [inline packages/node/example/multiSession/src/serverSideApp.ts:114]:
```suggestion
      .write(`<p>Logged out</p>`);
```

File: packages/node/example/multiSession/src/serverSideApp.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  } else {
+    res
+      .writeHead(400, { "Content-Type": "text/html" })
+      .write(`<p>No active session to log out</p>`);
+  }
+  res.end();
+});
+
+app.listen(PORT, async () => {
+  console.log(`Listening to ${PORT}`);
```
Comment (pmcb55) [inline packages/node/example/multiSession/src/serverSideApp.ts:124]:
```suggestion
  console.log(`Listening on [${PORT}}...`);
```

**Review by pmcb55 — APPROVED** [review]
Yeah, this looks great. Could use a bit more spit-and-polish, but proves the point nicely (i.e. we can get rid of the SessionManager).

File: packages/node/example/multiSession/src/serverSideApp.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      .writeHead(200, { "Content-Type": "text/html" })
+      .write(`<p>Alredy logged in</p>`);
+    res.end();
+  }
+});
+
+app.get("/redirect", async (req, res) => {
+  const session =
+    (await getStoredSession(req.session!.sessionId, storage)) ??
+    new Session({ storage }, req.session!.sessionId);
```
Comment (Vinnl) [inline packages/node/example/multiSession/src/serverSideApp.ts:69]: Does initialising a new session here work, or should we just return an error?

**Review by Vinnl — APPROVED** [review]
Looks good!

File: packages/node/example/multiSession/src/serverSideApp.ts (hunk 10 above)
Comment (NSeydoux) [reply packages/node/example/multiSession/src/serverSideApp.ts:69]: That's a good point: if there isn't information available in storage, the redirect can't succeed.

**Comment by NSeydoux** [comment]
Replaced with https://github.com/inrupt/solid-client-authn-js/pull/976

**Pull request closed without merging**